lib/iris/experimental/ugrid.py

```diff
@@ -2900,37 +2905,66 @@ def _construct_access_arrays(self):
         Mesh, according to the location and axis.
 
         Returns:
-        * points, bounds (array or None)
+        * points, bounds (array or None):
             lazy arrays which calculate the correct points and bounds from the
             Mesh data, based on the location and axis.
             The Mesh coordinates accessed are not identified on construction,
             but discovered from the Mesh at the time of calculation, so that
             the result is always based on current content in the Mesh.
 
         """
-        # TODO: cheat for now : correct calculations, but not dynamic.
-        # TODO: replace wth fully dynamic lazy calcs (slightly more complex).
         mesh, location, axis = self.mesh, self.location, self.axis
-        # N.B. mesh.coord returns a dict
-        node_coords = self.mesh.coord(include_nodes=True, axis=axis)
-        (node_coord,) = list(node_coords.values())
+        node_coord = _one_meshcoord(mesh.coord(include_nodes=True, axis=axis))
+
+        points_coord = None
+        bounds_connectivity = None
         if location == "node":
-            points = node_coord.core_points()
-            bounds = None
+            points_coord = node_coord
         elif location == "edge":
-            # N.B. mesh.coord returns a dict
-            edge_coords = self.mesh.coord(include_edges=True, axis=self.axis)
-            (edge_coord,) = list(edge_coords.values())
-            points = edge_coord.core_points()
-            inds = mesh.edge_node_connectivity.core_indices()
-            bounds = node_coord.core_points()[inds]
+            points_coord = _one_meshcoord(
+                mesh.coord(include_edges=True, axis=axis)
+            )
+            bounds_connectivity = mesh.edge_node_connectivity
         elif location == "face":
-            # N.B. mesh.coord returns a dict
-            face_coords = self.mesh.coord(include_faces=True, axis=self.axis)
-            (face_coord,) = list(face_coords.values())
-            points = face_coord.core_points()  # For now, this always exists
-            inds = mesh.face_node_connectivity.core_indices()
-            bounds = node_coord.core_points()[inds]
+            points_coord = _one_meshcoord(
+                mesh.coord(include_faces=True, axis=axis)
+            )
+            bounds_connectivity = mesh.face_node_connectivity
+
+        # The points output is the points of the relevant element-type coord.
+        points = points_coord.core_points()
+        if bounds_connectivity is None:
+            bounds = None
+        else:
+            # Bounds are calculated from a connectivity and the node points.
+            # Data can be real or lazy, so operations must work in Dask, too.
+            indices = bounds_connectivity.core_indices()
```

Comment:
```suggestion
            indices = bounds_connectivity.indices_by_src(bounds_connectivity.core_indices())
```

This functionality in `Connectivity` will allow us to handle transposed indices if present (`src_dim=1`), and without needing to modify the `Connectivity` instance itself, so it won't jeopardise future round-tripping.

Reply:
:+1: Dead right!
And it proves that `indices_by_src` really was a useful thing to provide.

Reply:
Actually, no I don't think we can use this because it is explicitly realising.
So, I **_think_** `connectivity.indices_by_src(self)` [--> self._indices](https://github.com/SciTools/iris/blob/be62cdfccd54e992ee8240a37ddea79c37772d6e/lib/iris/experimental/ugrid.py#L401), 
which then [--> self._values](https://github.com/SciTools/iris/blob/be62cdfccd54e992ee8240a37ddea79c37772d6e/lib/iris/experimental/ugrid.py#L380), 
[--> self._dm_data](https://github.com/SciTools/iris/blob/be62cdfccd54e992ee8240a37ddea79c37772d6e/lib/iris/coords.py#L228)
which [explicitly realises](https://github.com/SciTools/iris/blob/be62cdfccd54e992ee8240a37ddea79c37772d6e/lib/iris/_data_manager.py#L193-L204).

So, we really needed a `core_indices_by_src` or `lazy_indices_by_src` ?

Reply:
I've now added this, but put in a separate line of comment with a comment.

Reply:
P.S. now adding a test for this ...